packages/@glimmer/runtime/lib/compiled/opcodes/component.ts

```diff
@@ -174,12 +180,26 @@ APPEND_OPCODES.add(Op.ResolveDynamicComponent, (vm, { op1: _meta }) => {
     let resolvedDefinition = resolveComponent(vm.runtime.resolver, component, meta);
 
     definition = expect(resolvedDefinition, `Could not find a component named "${component}"`);
-  } else if (isCurriedComponentDefinition(component)) {
-    definition = component;
   } else {
-    throw unreachable();
+    definition = component;
+  }
+
+  stack.pushJs(definition);
+});
+
+APPEND_OPCODES.add(Op.ResolveCurriedComponent, (vm) => {
+  let stack = vm.stack;
+  let value = valueForRef(check(stack.popJs(), CheckReference));
+
+  if (DEBUG && !isCurriedComponentDefinition(value)) {
+    throw new Error(
+      `A curried component definition. You may have accidentally done <this.dynamicComponent>, where dynamicComponent was a string instead of a curried component definition. You must use the {{component}} helper to create a component definition when invoking dynamically.`
```

Comment:
is it possible to use real path / value for error?

one more use case is:

```hbs
{{#let "my-component" as |AnyComponent|}}
   <AnyComponent />
{{/let}}
```

Reply:
Updated the error message to reference the debugLabel of the reference, which should give better context in general.

Re: the use case, that is what we are trying to prevent, and is not a valid use case in general. You should convert forward to:

```hbs
{{#let (hash AnyComponent=(component "my-component")) as |ctx|}}
   <ctx.AnyComponent />
{{/let}}
```

Reply:
I know it not valid, my comment point is to have debug information pointing to such usage :)
(error may appear not only in `this.foo` invocation)